### PR TITLE
[posix] add multicast address update in netif module

### PR DESCRIPTION
**Pull request by Irving-cl**
This PR addes multicast addresses update method of `Netif` module.

This PR contains a small unit test case to verify the implementation.

### CODE REVIEW

File: src/ncp/posix/netif.cpp [hunk 1]
```diff
@@ -114,6 +114,65 @@ void Netif::UpdateIp6UnicastAddresses(const std::vector<Ip6AddressInfo> &aAddrIn
...
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    // Remove stale addresses
+    for (const Ip6Address &address : mIp6MulticastAddresses)
+    {
+        if (std::find(aAddrs.begin(), aAddrs.end(), address) == aAddrs.end())
+        {
+            otbrLogInfo("Remove address: %s", Ip6Address(address).ToString().c_str());
+            SuccessOrExit(error = ProcessMulticastAddressChange(address, false));
```
Comment (abtink) [inline src/ncp/posix/netif.cpp:127]: Add comment for boolean arg? 
```suggestion
            SuccessOrExit(error = ProcessMulticastAddressChange(address, /* aIsAdded */ false));
```

File: src/ncp/posix/netif.cpp [hunk 2]
```diff
@@ -114,6 +114,65 @@ void Netif::UpdateIp6UnicastAddresses(const std::vector<Ip6AddressInfo> &aAddrIn
...
+    }
+
+    // Add new addresses
+    for (const Ip6Address &address : aAddrs)
+    {
+        if (std::find(mIp6MulticastAddresses.begin(), mIp6MulticastAddresses.end(), address) ==
+            mIp6MulticastAddresses.end())
+        {
+            otbrLogInfo("Add address: %s", Ip6Address(address).ToString().c_str());
+            SuccessOrExit(error = ProcessMulticastAddressChange(address, true));
```
Comment (abtink) [inline src/ncp/posix/netif.cpp:138]:
```suggestion
            SuccessOrExit(error = ProcessMulticastAddressChange(address, /* aIsAdded */ true));
```

**Review by abtink — APPROVED** [review]
LGTM. Thanks. 👍 
Two smaller (optional) suggestions below.

File: src/ncp/posix/netif.cpp [hunk 3]
```diff
@@ -114,6 +114,65 @@ void Netif::UpdateIp6UnicastAddresses(const std::vector<Ip6AddressInfo> &aAddrIn
...
+    return error;
+}
+
+otbrError Netif::ProcessMulticastAddressChange(const Ip6Address &aAddress, bool aIsAdded)
+{
+    struct ipv6_mreq mreq;
+    otbrError        error = OTBR_ERROR_NONE;
+    int              err;
+
+    VerifyOrExit(mIpFd >= 0);
```
Comment (sunytt) [inline src/ncp/posix/netif.cpp:158]: mIpFd < 0 seems to be a serious problem, shall we return an error for it or fail harder?

File: tests/gtest/test_netif.cpp [hunk 4]
```diff
@@ -297,4 +360,55 @@ TEST(Netif, WpanIfHasCorrectUnicastAddresses_AfterUpdatingUnicastAddresses)
...
+    otbr::Ip6Address kMulAddr2 = {
+        {0xff, 0x32, 0x00, 0x40, 0xfd, 0x0d, 0x07, 0xfc, 0xa1, 0xb9, 0xf0, 0x50, 0x00, 0x00, 0x00, 0x01}};
+    const char *kMulAddr1Str = "ff03::fc";
+    const char *kMulAddr2Str = "ff32:40:fd0d:7fc:a1b9:f050:0:1";
+
+    otbr::Ip6Address testArray1[] = {
+        kMulAddr1,
+    };
+    std::vector<otbr::Ip6Address> testVec1(testArray1, testArray1 + sizeof(testArray1) / sizeof(otbr::Ip6Address));
+    netif.UpdateIp6MulticastAddresses(testVec1);
```
Comment (sunytt) [inline tests/gtest/test_netif.cpp:385]: may also test the case that netif.UpdateIp6MulticastAddresses contains one of the default subscribed multicast addresses?

File: src/ncp/posix/netif.cpp (hunk 3 above)
Comment (Irving-cl) [reply src/ncp/posix/netif.cpp]: Returned 'INVALID_STATE' for it.

File: tests/gtest/test_netif.cpp (hunk 4 above)
Comment (Irving-cl) [reply tests/gtest/test_netif.cpp:387]: Sounds good! Added a case for that.

File: tests/gtest/test_netif.cpp [hunk 5]
```diff
@@ -297,4 +360,65 @@ TEST(Netif, WpanIfHasCorrectUnicastAddresses_AfterUpdatingUnicastAddresses)
...
+    EXPECT_THAT(wpanMulAddrs, ::testing::Contains(kDefaultMulAddr2Str));
+    EXPECT_THAT(wpanMulAddrs, ::testing::Contains(kDefaultMulAddr3Str));
+
+    otbr::Ip6Address              testArray3[] = {kDefaultMulAddr1};
+    std::vector<otbr::Ip6Address> testVec3(testArray3, testArray3 + sizeof(testArray3) / sizeof(otbr::Ip6Address));
+    netif.UpdateIp6MulticastAddresses(testVec3);
+    EXPECT_EQ(wpanMulAddrs.size(), 3);
+    EXPECT_THAT(wpanMulAddrs, ::testing::Contains(kDefaultMulAddr1Str));
+    EXPECT_THAT(wpanMulAddrs, ::testing::Contains(kDefaultMulAddr2Str));
+    EXPECT_THAT(wpanMulAddrs, ::testing::Contains(kDefaultMulAddr3Str));
```
Comment (sunytt) [inline tests/gtest/test_netif.cpp:420]: Add another update to empty list to see the kDefaultMulAddr1 is still subscribed?

Comment (Irving-cl) [reply tests/gtest/test_netif.cpp:420]: Sounds good! I moved updating to empty to the end and verify `kDefaultMulAddr1`is still subscribed. This is equivalent.

**Review by sunytt — APPROVED** [review]
LGTM 👍

**Review by jwhui — APPROVED** [review]

**Pull request merged**